api/src/main/java/org/apache/iceberg/metrics/FixedReservoirHistogram.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import java.util.Arrays;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A {@link Histogram} implementation that uses a fixed size reservoir like circular buffer. Oldest
+ * observation is replaced by latest observation.
+ */
+public class FixedReservoirHistogram implements Histogram {
+  private final long[] measurements;
+  private long count;
+
+  public FixedReservoirHistogram(int reservoirSize) {
+    this.measurements = new long[reservoirSize];
+    this.count = 0L;
+  }
+
+  @Override
+  public synchronized long count() {
+    return count;
+  }
+
+  @Override
+  public synchronized void update(long value) {
+    int index = (int) (count % measurements.length);
+    measurements[index] = value;
+    count++;
+  }
+
+  @Override
+  public Statistics statistics() {
+    int size = measurements.length;
+    if (count < (long) measurements.length) {
+      size = (int) count;
+    }
+
+    long[] values = new long[size];
+    for (int i = 0; i < values.length; ++i) {
+      synchronized (this) {
+        values[i] = measurements[i];
+      }
+    }
+
+    return new UniformWeightStatistics(values);
+  }
+
+  private static class UniformWeightStatistics implements Statistics {
+    private final long[] values;
+
+    private UniformWeightStatistics(long[] values) {
```

Comment:
Rather than copying all of the values to a new array and lazily computing stats, wouldn't it make more sense to compute stats over the sampled values by iterating here once?

Right now, if you want the stddev and mean, this would:
1. Allocate a new array of length `L = min(measurements.length, count)`
2. Copy `L` measurements into the new array
3. Iterate over `L` measurements to produce a mean (to return)
4. Iterate over `L` measurements to produce a mean (for stddev calculation)
5. Iterate over `L` measurements to produce `sumOfDiffSquares`

I think it would be a lot better to calculate all of the stats in one pass over the values in the constructor:

```java
  private UniformWeightStatistics(long[] measurements, int count) {
    double sumSquares = 0.0d;
    double sum = 0.0d;
    this.min = Long.MAX_VALUE;
    this.max = Long.MIN_VALUE;
    for (int i = 0; i < count; i += 1) {
      sumSquares += measurements[i] * measurements[i];
      sum += measurements[i];
      if (measurements[i] < min) {
        this.min = measurements[i];
      }
      if (measurements[i] > max) {
        this.max = measurements[i];
      }
    }
    this.mean = sum / count;
    this.stddev = Math.sqrt((sumSquares / count) - (mean * mean));
  }
```

Reply:
Ah, I missed that you're doing the sort on the values in the copy. It's probably fine to copy still, but you can calculate the mean and stddev while you copy, before sorting the values.

Reply:
yeah. avoiding multiple loops makes sense.

Reply:
Looking at the standard deviation again, I don't think `Math.sqrt((sumSquares / count) - (mean * mean))` is correct. 

We need to calculate the mean first in order to compute `sumOfDiffSquares`. But we can avoid computing the mean twice.

Reply:
The `sumSquares` approach is correct and equivalent. It is the ["Naive algorithm" described on Wikipedia](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance). We should probably also apply Bessel's correction as well since this is a sample (as noted on that page).

There's an [alternative algorithm](https://math.stackexchange.com/questions/198336/how-to-calculate-standard-deviation-with-streaming-inputs) mentioned on stack exchange, too. I've always used the naive one, since it's easy.

Reply:
Thanks for the wikipedia link. I updated the impl.

I didn't apply Bessel's correction. According to the [caveats section](https://en.wikipedia.org/wiki/Bessel%27s_correction#Caveats), it is probably not the right fit here. Also the Bessel's correction can't work with `n == 1`.